### PR TITLE
Dynamic Pagination in CPT Dashboard

### PR DESCRIPTION
**Pull request by MVarshini**
## Type of change

- [ ] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] Optimization
- [ ] Documentation Update

## Description

The `[PROD]_JOBS_API` accepts `size`  and `offset`  params 

## Related Tickets & Documents

- Related Issue # [PANDA-651](https://issues.redhat.com/browse/PANDA-651)
- Closes #

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.

## Testing
- Please describe the System Under Test.
- Please provide detailed steps to perform tests related to this code change.
- How were the fix/results from this change verified? Please provide relevant screenshots or results.

### CODE REVIEW

File: backend/app/api/v1/endpoints/ocm/ocmJobs.py
Comment (dbutenhof) [inline backend/app/api/v1/endpoints/ocm/ocmJobs.py:24]: s/contet/content/ -- I think you copy-pasted this a lot.

File: backend/app/api/v1/endpoints/ocp/ocpJobs.py
Comment (dbutenhof) [inline backend/app/api/v1/endpoints/ocp/ocpJobs.py:15]: I think you should remove the "of requested size" ... the `size` and `offset` query parameters are self-describing, and this makes the sentence unnecessarily awkward.

**Review by dbutenhof — COMMENTED** [review]
I didn't get all the way through, but here's a bunch ...

File: backend/app/api/v1/endpoints/cpt/maps/hce.py
Comment (dbutenhof) [reply backend/app/api/v1/endpoints/cpt/maps/hce.py:29]: And, as a general maintainability/readability suggestion to the maintainers: after all the outstanding PRs are merged and stabilized, it'd be a great idea to
1. run `black` and `Prettier` on all Python and JavaScript/CSS files, and
2. add a CI lint check GitHub Action that'll fail a PR if it breaks the style rules. (Of course adding additional lint checks with `flake8`/ES lint/CodeQL would be a great idea too.)
3. Submit this as a PR with no *functional* changes; it might be large, (depending on how bad the code is), but it'd be straightforward to review.

File: backend/app/api/v1/commons/hce.py
Comment (vishnuchalla) [reply backend/app/api/v1/commons/hce.py:31]: @MVarshini maybe we can just have this check at one place from where this virtual function is being invoked?

File: frontend/src/components/organisms/Pagination/index.jsx
Comment (MVarshini) [reply frontend/src/components/organisms/Pagination/index.jsx]: Removed the commented line.

File: backend/app/api/v1/endpoints/ocp/ocpJobs.py
Comment (MVarshini) [reply backend/app/api/v1/endpoints/ocp/ocpJobs.py]: Removed the unnecessary phrase.

File: backend/app/api/v1/commons/hce.py
Comment (MVarshini) [reply backend/app/api/v1/commons/hce.py]: okay

File: backend/app/api/v1/commons/hce.py
Comment (MVarshini) [reply backend/app/api/v1/commons/hce.py:8]: Got it!

File: backend/app/api/v1/commons/hce.py
Comment (MVarshini) [reply backend/app/api/v1/commons/hce.py]: Updated the non-empty case too.

File: backend/app/services/search.py
Comment (MVarshini) [reply backend/app/services/search.py:122]: I'm concerned that this change might inadvertently impact other scenarios, even though I've verified it against most cases. If it doesn't affect any other cases then these lines can be taken out or it can be reverted.

File: backend/app/api/v1/commons/ocp.py
Comment (dbutenhof) [inline backend/app/api/v1/commons/ocp.py:30]: I think that `'data':jbs` is a typo. There isn't actually a `jbs` in scope, is there?

File: backend/app/api/v1/commons/quay.py
Comment (dbutenhof) [inline backend/app/api/v1/commons/quay.py:30]: `jbs` => `jobs`

File: backend/app/api/v1/endpoints/ocp/ocpJobs.py
Comment (dbutenhof) [inline backend/app/api/v1/endpoints/ocp/ocpJobs.py]: Maybe "in the specified date range"?

**Review by dbutenhof — CHANGES_REQUESTED** [review]
Lots of good cleanup, but I think you added a few bugs while you were at it. 😆

File: backend/app/api/v1/commons/hce.py
Comment (jaredoconnell) [reply backend/app/api/v1/commons/hce.py]: Yeah, is this necessary? Is the response value `total` not present when the number of jobs is zero?

File: backend/app/api/v1/commons/quay.py
Comment (MVarshini) [reply backend/app/api/v1/commons/quay.py:30]: Actually, 

`  cleanJobs = jobs[jobs["platform"] != ""]  
    jbs = cleanJobs `
    
    It's not typo :smirk:

File: backend/app/api/v1/endpoints/cpt/maps/hce.py
Comment (dbutenhof) [reply backend/app/api/v1/endpoints/cpt/maps/hce.py]: I hope that rebasing all of the outstanding PRs against the formatting changes in #139 won't cause major pain ... but it'll be nice to have the reformatting out of the way. I plan to add `black --check` to the CI; probably along with `flake8`.

**Comment by dbutenhof** [comment]
And now that the reformatting is merged ... you've got a bunch of conflicts to resolve here. 😆 

Once you've rebased and resolved the conflicts, make sure that you run `black` again on the results to be safe!

**Review by dbutenhof — COMMENTED** [review]
And now that you've merged the `black` reformatting, I recommend that you make a habit of running `black` before pushing any commit to a PR. At some point we'll be able to turn on `black --check` in the CI so we won't accidentally break the conventions -- but even then, doing it proactively is a good idea. (You can turn on git pre-commit hooks if you want, though I've found that can sometimes become annoying and I prefer to do it manually when I'm ready.)

File: backend/app/api/v1/commons/utils.py
Comment (MVarshini) [reply backend/app/api/v1/commons/utils.py:11]: It's legacy code and I didn't touch it.

File: backend/app/api/v1/commons/quay.py
Comment (MVarshini) [reply backend/app/api/v1/commons/quay.py]: "`total`"  should be `response["total"]`. `len(cleanJobs)`  is just the size requested in the API whereas `response["total"]`  is the total number of jobs within the specified date range.

File: backend/app/api/v1/endpoints/ocp/ocpJobs.py
Comment (MVarshini) [reply backend/app/api/v1/endpoints/ocp/ocpJobs.py]: Updated!

File: backend/app/services/search.py
Comment (MVarshini) [reply backend/app/services/search.py]: Removed them :laughing:

**Review by dbutenhof — COMMENTED** [review]
A few new comments, but they're more style suggestions; and one remaining unanswered confusion about your logic at https://github.com/cloud-bulldozer/cpt-dashboard/pull/135#discussion_r1876009762; but this is looking good.

File: backend/app/api/v1/endpoints/ocm/ocmJobs.py
Comment (jaredoconnell) [reply backend/app/api/v1/endpoints/ocm/ocmJobs.py:62]: Dave's suggestion would definitely be clearer.

**Review by jaredoconnell — CHANGES_REQUESTED** [review]

File: backend/app/api/v1/commons/hce.py
Comment (dbutenhof) [reply backend/app/api/v1/commons/hce.py:28]: There are a lot of potential refactoring ideas we should pursue to make the somewhat messy codebase more maintainable.

Right now, we've got a long list of PRs we'd really like to land, and we need to focus on the most critical concerns ... starting with landing this PR, #138, and then a follow-on filtering PR which ought to complete the "revamp" branch. Once we land that onto "main" we can churn through my backlog of ilab PRs.

At that point, our focus should be cleaning up the code base (including unit testing, functional testing, lint and format checkers) to make it more maintainable with a viable CI.

As much as I hate being "pragmatic" in things like this, let's not let minor stuff like this bog us down at this point. Maybe leaving these comments in place and open as a reference for later work isn't a bad idea ...

File: backend/app/api/v1/commons/ocm.py
Comment (dbutenhof) [reply backend/app/api/v1/commons/ocm.py:38]: I don't find long conditional blocks necessarily easier to follow. In the former case there was the one-line early return vs 2 indented conditional lines for a common return, and I think that's a good call. This block is getting big enough that the early return might be more readable... although there are probably much better ways to refactor the method to avoid this sort of thing in the future.

File: backend/app/api/v1/commons/ocp.py
Comment (dbutenhof) [reply backend/app/api/v1/commons/ocp.py:60]: `cleanJobs` is the result of a pandas `dataframe` filter to remove rows, although my understanding of the complicated pandas infrastructure is minimal.

What bugs me here is the `jbs = cleanJobs` to effectively just obscure the name before returning the data using `jbs`. That's not new with Varshini's changes, though, and I don't think it's practical to push her to rewrite more existing logic than necessary for the revamp. (On the other hand, simply replacing `cleanJobs` with `jbs` or dropping the intermediary `jbs` *would* make the code easier to read, and it'd be "simple"... it's just that an endless list of "simple" things isn't simple anymore. 🤔 💣 )

File: backend/app/api/v1/endpoints/cpt/cptJobs.py
Comment (dbutenhof) [reply backend/app/api/v1/endpoints/cpt/cptJobs.py:94]: We decided to flip this by overriding `total` with the incoming `totalJobs` rather than modifying the query parameter, and adding a comment to explain the logic.

File: backend/app/api/v1/endpoints/cpt/maps/hce.py
Comment (dbutenhof) [reply backend/app/api/v1/endpoints/cpt/maps/hce.py:19]: Yeah, the import should be at the top. I'm wondering whether `isort` (which I intent to run when we move into code ownership/cleanup after our PRs are merged) will fix this??? 🤔 In any case, it's a new `import` and ought to be moved up to the top section.

File: frontend/src/actions/paginationActions.js
Comment (dbutenhof) [reply frontend/src/actions/paginationActions.js:80]: Because each tab/component does really different things, and therefore the "common" components have distinct cases. It's not a great design. For example, we might be better off providing common base classes that get extended by components to handle behavioral differences ... assuming those differences can't simply be resolved.

For now, although the code has been pushed and stretched a bit around the edges, this aspect of the logic isn't new and we really need to push these PRs through and move on so let's defer the analysis of possible deep architectural cleanup.

**Review by dbutenhof — COMMENTED** [review]
I think this is really close, with just a few minor cleanups needed to push it forward so that Varshini can concentrate on finishing the `revamp` branch and letting us move on.

File: backend/app/api/v1/endpoints/cpt/maps/hce.py
Comment (MVarshini) [reply backend/app/api/v1/endpoints/cpt/maps/hce.py]: Moved the imports to the top

File: backend/app/api/v1/endpoints/cpt/maps/hce.py
Comment (jaredoconnell) [inline backend/app/api/v1/endpoints/cpt/maps/hce.py:20]: Now that the import is removed, should this gap still be here?

File: backend/app/api/v1/commons/ocp.py
Comment (jaredoconnell) [reply backend/app/api/v1/commons/ocp.py:60]: We will definitely need to keep a note of this for a future refactor PR.

**Review by jaredoconnell — APPROVED** [review]
I just have one new comment. The rest can be addressed later.

**Comment by chentex** [comment]
I tested the Pr locally and gone through the changes, I think we are good to go, let's wait also for @vishnuchalla to take an extra look and test.

**Review by chentex — APPROVED** [review]

**Comment by vishnuchalla** [comment]
Few observations while navigating through the dashboard.
* When a filter is being applied the row count is being updated but the table rows are not being filtered. Applies for all the tabs. For example, try filtering out using benchmark. This seems like an issue while slicing out the rows in frontend code.
* Graph data is not being rendered in Quay tab (i.e Error's out in ES query). Works for the rest of the tabs.
* Pagination works properly only for `next` page but not when trying to navigate to `previous`, `first` and `last` pages. (i.e. the rows in the table are not getting updated at all). After looking at the logs, this seems like there is not query being sent from frontend to the backend with updated `size/offset` on a state update.

**Comment by vishnuchalla** [comment]
> Few observations while navigating through the dashboard.
> 
> * When a filter is being applied the row count is being updated but the table rows are not being filtered. Applies for all the tabs. For example, try filtering out using benchmark. This seems like an issue while slicing out the rows in frontend code.
> * Graph data is not being rendered in Quay tab (i.e Error's out in ES query). Works for the rest of the tabs.
> * Pagination works properly only for `next` page but not when trying to navigate to `previous`, `first` and `last` pages. (i.e. the rows in the table are not getting updated at all). After looking at the logs, this seems like there is not query being sent from frontend to the backend with updated `size/offset` on a state update.
> 
> cc: @dbutenhof @MVarshini @jaredoconnell @chentex @jtaleric

If we want to take these improvements/bug fixes in follow up PRs and get us unblocked for now, we can get this PR merged into `revamp` branch. Merging this into staging will introduce bugs and block others from making their changes and promoting things to production. Open for other thoughts too.

**Comment by dbutenhof** [comment]
Thanks for the review and testing, @vishnuchalla. (Now we need to encapsulate all that testing into an automated CI ... but we'll get there.)

@MVarshini already has one more PR lined up on the `revamp` branch for additional issues, plus another branch she's working on now that doesn't yet have a PR. I think we need to get "revamp" to product quality so we can drop it onto main, cut a release, and move on.

All three of the issues you found need to be fixed on revamp before we drop to main and stage it, but not necessarily in this PR unless we can point out changed code that's responsible. Since #138 focuses on backend sorting, it's likely to touch these issues only by accident, but I don't remember the exact scope of the remaining PR so it's possible there's some overlap already?

**Comment by dbutenhof** [comment]
I created new issues to record @vishnuchalla's three problems on the REVAMP branch. We should re-evaluate these after @MVarshini's remaining two PRs are stabilized, and fix them if still necessary before dropping `revamp` onto `main` for staging.

I'm going to go ahead and merge this PR so she can move forward with #135.

**Pull request merged**
